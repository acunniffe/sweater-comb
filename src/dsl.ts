import {
  ApiCheckDsl,
  EntityRule,
  Result,
  DocsLinkHelper,
  newDocsLinkHelper,
  runCheck,
} from "@useoptic/api-checks";
import {
  ConceptualLocation,
  OpenApiHeaderFact,
  OpenApiKind,
  OpenApiOperationFact,
  IChange,
  IFact,
  OpenApiFieldFact,
  jsonPointerHelper,
  ILocation,
  OpenAPIV3,
} from "@useoptic/openapi-utilities";
import { genericEntityRuleImpl } from "@useoptic/api-checks/build/sdk/generic-entity-rule-impl";
import { ShouldOrMust } from "@useoptic/api-checks/build/sdk/types";
import { OpenApiResponseFact } from "@useoptic/openapi-utilities/build/openapi3/implementations/openapi3/openapi-traverser";

type SnykStablity = "wip" | "experimental" | "beta" | "ga";
type DateString = string; // YYYY-mm-dd
type ResourceName = string;

export interface SynkApiCheckContext {
  // Vervet provides context about the change itself. Since
  // Optic is analyzing two OpenAPI spec files, we need to tell it
  // when the change is supposed to happen, and the resource/version info
  // determined by the file's location in a directory structure.
  changeDate: DateString; // when the change did (or would, if proposed) occur
  changeResource: ResourceName; // the spec resource being changed
  changeVersion: {
    // the spec version being changed
    date: DateString;
    stability: SnykStablity;
  };

  // Vervet provides a mapping that indicates the resource version deprecation.
  // It has this information because it processes the entire source tree of
  // spec files.
  resourceVersions: {
    [ResourceName: string]: {
      // changeResource used to match this
      [DateString: string]: {
        // changeVersion.date used to match this
        [SnykStablity: string]: {
          // changeVersion.stability matches this
          deprecatedBy: {
            // the spec version that deprecates this one (if any) or null
            date: DateString;
            stability: SnykStablity; // could be higher stability than changed!
          } | null;
        };
      };
    };
  };
}

export interface SnykEntityRule<T, A>
  extends EntityRule<T, ConceptualLocation, SynkApiCheckContext, A> {}

export class SnykApiCheckDsl implements ApiCheckDsl {
  private checks: Promise<Result>[] = [];

  constructor(
    private nextFacts: IFact<any>[],
    private changelog: IChange<any>[],
    private nextJsonLike: OpenAPIV3.Document,
    private providedContext: SynkApiCheckContext
  ) {}

  checkPromises() {
    return this.checks;
  }

  getContext(location: ILocation): ConceptualLocation & SynkApiCheckContext {
    return {
      ...location.conceptualLocation,
      ...this.providedContext,
    };
  }

  get operations() {
    return genericEntityRuleImpl<
      OpenApiOperationFact,
      ConceptualLocation,
      SynkApiCheckContext,
      OpenAPIV3.OperationObject
    >(
      OpenApiKind.Operation,
      this.changelog,
      this.nextFacts,
      (opFact) => `${opFact.method.toUpperCase()} ${opFact.pathPattern}`,
      (location) => this.getContext(location),
      (...items) => this.checks.push(...items),
      (pointer: string) => jsonPointerHelper.get(this.nextJsonLike, pointer)
    );
  }

  get context() {
    const change: IChange<SnykApiCheckDsl> = {
      location: {
        conceptualLocation: { path: "This Specification", method: "" },
        jsonPath: "/",
        conceptualPath: [],
        kind: "API",
      },
    };

    const value: ShouldOrMust<
      (
        context: SynkApiCheckContext,
        docs: DocsLinkHelper
      ) => Promise<void> | void
    > = {
      must: (statement, handler) => {
        const docsHelper = newDocsLinkHelper();
        return runCheck(
          change,
          docsHelper,
          "this specification: ",
          statement,
          true,
          () => handler(this.providedContext, docsHelper)
        );
      },
      should: (statement, handler) => {
        const docsHelper = newDocsLinkHelper();
        return runCheck(
          change,
          docsHelper,
          "this specification: ",
          statement,
          false,
          () => handler(this.providedContext, docsHelper)
        );
      },
    };

    return value;
  }

  get responses() {
    const dsl = this;

    return {
      ...genericEntityRuleImpl<
        OpenApiResponseFact,
        ConceptualLocation,
        SynkApiCheckContext,
        OpenAPIV3.ResponsesObject
      >(
        OpenApiKind.Response,
        dsl.changelog,
        dsl.nextFacts,
        (response) => `response ${response.statusCode}`,
        (location) => dsl.getContext(location),
        (...items) => dsl.checks.push(...items),
        (pointer: string) => jsonPointerHelper.get(dsl.nextJsonLike, pointer)
      ),
      headers: genericEntityRuleImpl<
        OpenApiHeaderFact,
        ConceptualLocation,
        SynkApiCheckContext,
        OpenAPIV3.HeaderObject
      >(
        OpenApiKind.ResponseHeader,
        dsl.changelog,
        dsl.nextFacts,
        (header) => `response header ${header.name}`,
        (location) => dsl.getContext(location),
        (...items) => dsl.checks.push(...items),
        (pointer: string) => jsonPointerHelper.get(dsl.nextJsonLike, pointer)
      ),
    };
  }

  get checkApiContext(): ShouldOrMust<
    (context: SynkApiCheckContext, docs: DocsLinkHelper) => void
  > {
    const contextChangedHandler: (must: boolean) => ContextChangedRule["must"] =
      (must: boolean) => {
        return (statement, handler) => {
          const docsHelper = newDocsLinkHelper();
          const syntheticChange = {
            added: this.providedContext,
            location: {
              jsonPath: "/",
              conceptualPath: [],
              conceptualLocation: {
                path: "Entire Resource",
                method: "",
              },
              kind: "ContextRule",
            },
          };
          return runCheck(
            syntheticChange,
            docsHelper,
            "api lifeycle: ",
            statement,
            must,
            () => handler(this.providedContext, docsHelper)
          );
        };
      };

    return {
      must: contextChangedHandler(true),
      should: contextChangedHandler(false),
    };
  }

  get bodyProperties(): SnykEntityRule<
    OpenApiFieldFact,
    OpenAPIV3.SchemaObject
  > {
    const dsl = this;
    return genericEntityRuleImpl<
      OpenApiFieldFact,
      ConceptualLocation,
      SynkApiCheckContext,
      OpenAPIV3.SchemaObject
    >(
      OpenApiKind.Field,
      dsl.changelog,
      dsl.nextFacts,
      (field) => `field ${field.key}`,
      (location) => dsl.getContext(location),
      (...items) => dsl.checks.push(...items),
      (pointer: string) => jsonPointerHelper.get(dsl.nextJsonLike, pointer)
    );
  }
}

type ContextChangedRule = ShouldOrMust<
  (context: SynkApiCheckContext, docs: DocsLinkHelper) => void
>;
