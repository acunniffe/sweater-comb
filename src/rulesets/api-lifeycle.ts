import { SnykApiCheckDsl } from "../dsl";
import { expect } from "chai";

export const rules = {
  example: ({ checkApiContext, responses }: SnykApiCheckDsl) => {
    checkApiContext.must(
      "lifeycle rules have to be followed",
      (context, docs) => {
        docs.includeDocsLink("https://how.we.version/rule");
        context.changeVersion.date;
      },
    );
  },

  stabilityRequirement: ({ stability }: SnykApiCheckDsl) => {
    stability.must(
      "be provided for every resource document",
      (before, after, context, docs) => {
        const allowed = ["wip", "experimental", "beta", "ga"];
        const wasValid = allowed.includes(after || "");

        expect(wasValid, `${after} must be one of allowed values ${allowed}`).to
          .be.ok;
      },
    );
  },

  allowedStabilityTransitions: ({ stability }: SnykApiCheckDsl) => {
    stability.must(
      "not change unless it was wip",
      (before, after, context, docs) => {
        if (!before && !after) return;
        // Allow for creating new resource versions with any stability.
        // No `before` means this is a new resource
        if (!before) return;
        if (before === "wip") return;
        expect(before).to.equal(after);
      },
    );
  },
};
