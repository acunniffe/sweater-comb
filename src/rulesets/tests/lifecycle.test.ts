import { rules } from "../api-lifeycle";
import { SnykApiCheckDsl, SynkApiCheckContext } from "../../dsl";
import { createSnykTestFixture } from "./fixtures";

const { compare } = createSnykTestFixture();

const emptyContext: SynkApiCheckContext = {
  changeDate: "2021-10-10",
  changeResource: "Example",
  changeVersion: {
    date: "2021-10-10",
    stability: "ga",
  },
  resourceVersions: {},
};

describe("lifecycle", () => {
  // todo: fix copy/paste
  const baseOpenAPI = {
    openapi: "3.0.1",
    paths: {
      "/example": {
        get: {
          responses: {},
        },
      },
    },
    info: { version: "0.0.0", title: "OpenAPI" },
  };

  const withWip = {
    openapi: "3.0.1",
    "x-snyk-api-stability": "wip",
    paths: {
      "/example": {
        get: {
          responses: {},
        },
      },
    },
    info: { version: "0.0.0", title: "OpenAPI" },
  };
  const withGa = {
    openapi: "3.0.1",
    "x-snyk-api-stability": "ga",
    paths: {
      "/example": {
        get: {
          responses: {},
        },
      },
    },
    info: { version: "0.0.0", title: "OpenAPI" },
  };

  describe("stability", () => {
    it("must be provided", async () => {
      const result = await compare(baseOpenAPI)
        .to((spec) => {
          return spec;
        })
        .withRule(rules.stabilityRequirement, emptyContext);

      expect(result.results.every((i) => i.passed)).toBeFalsy();
      expect(result).toMatchSnapshot();
    });

    it("must be one of the allowed values", async () => {
      const result = await compare(baseOpenAPI)
        .to((spec) => {
          spec["x-snyk-api-stability"] = "published";
          return spec;
        })
        .withRule(rules.stabilityRequirement, emptyContext);

      expect(result.results.every((i) => i.passed)).toBeFalsy();
      expect(result).toMatchSnapshot();
    });

    it("wip can change to another maturity", async () => {
      const result = await compare(withWip)
        .to((spec) => {
          spec["x-snyk-api-stability"] = "beta";
          return spec;
        })
        .withRule(rules.allowedStabilityTransitions, emptyContext);

      expect(result.results.every((i) => i.passed)).toBeTruthy();
      expect(result).toMatchSnapshot();
    });

    it("can not change from any stability but wip", async () => {
      const result = await compare(withGa)
        .to((spec) => {
          spec["x-snyk-api-stability"] = "beta";
          return spec;
        })
        .withRule(rules.allowedStabilityTransitions, emptyContext);

      expect(result.results.every((i) => i.passed)).toBeFalsy();
      expect(result).toMatchSnapshot();
    });
  });

  describe("sunset", () => {
    it("fails when the file was removed and not deprecated", async () => {
      const result = await compare(baseOpenAPI)
        .to((spec) => {
          spec["x-optic-ci-empty-spec"] = true;
          return spec;
        })
        .withRule(rules.followSunsetRules, emptyContext);
      expect(result.results[0].passed).toBeFalsy();
      expect(result).toMatchSnapshot();
    });

    describe("sunset schedule", () => {
      const context: SynkApiCheckContext = {
        changeDate: "2021-10-20",
        changeResource: "Example",
        changeVersion: {
          date: "2021-10-10",
          stability: "experimental",
        },
        resourceVersions: {
          Example: {
            "2021-10-10": {
              experimental: {
                deprecatedBy: {
                  date: "2021-10-20",
                  stability: "ga",
                },
              },
            },
          },
        },
      };

      it("fails when the schedule isn't met", async () => {
        const result = await compare(baseOpenAPI)
          .to((spec) => {
            spec["x-optic-ci-empty-spec"] = true;
            return spec;
          })
          .withRule(rules.followSunsetRules, context);
        expect(result.results[0].passed).toBeFalsy();
        expect(result).toMatchSnapshot();
      });

      it("passes when the schedule is met", async () => {
        const result = await compare(baseOpenAPI)
          .to((spec) => {
            spec["x-optic-ci-empty-spec"] = true;
            return spec;
          })
          .withRule(rules.followSunsetRules, {
            ...context,
            changeDate: "2021-12-01",
          });
        expect(result.results[0].passed).toBeTruthy();
        expect(result).toMatchSnapshot();
      });
    });
  });
});
