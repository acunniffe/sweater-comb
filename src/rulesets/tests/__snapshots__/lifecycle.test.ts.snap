// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lifecycle stability can not change from any stability but wip 1`] = `
Object {
  "base": Object {
    "info": Object {
      "title": "OpenAPI",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/example": Object {
        "get": Object {
          "responses": Object {},
        },
      },
    },
    "x-snyk-api-stability": "ga",
  },
  "changelog": Array [],
  "next": Object {
    "info": Object {
      "title": "OpenAPI",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/example": Object {
        "get": Object {
          "responses": Object {},
        },
      },
    },
    "x-snyk-api-stability": "beta",
  },
  "results": Array [
    Object {
      "change": Object {
        "changed": Object {
          "after": "beta",
          "before": "ga",
        },
        "location": Object {
          "conceptualLocation": Object {
            "method": "",
            "path": "Resource Document",
          },
          "conceptualPath": Array [
            "x-snyk-api-stability",
          ],
          "jsonPath": "/x-snyk-api-stability",
          "kind": "SnykStability",
        },
      },
      "condition": "not change unless it was wip",
      "docsLink": "https://github.com/snyk/sweater-comb/blob/main/docs/version.md#promoting-stability-of-a-resource-over-time",
      "error": "expected 'ga' to equal 'beta'",
      "isMust": true,
      "isShould": false,
      "passed": false,
      "where": "published stability: ",
    },
  ],
}
`;

exports[`lifecycle stability must be one of the allowed values 1`] = `
Object {
  "base": Object {
    "info": Object {
      "title": "OpenAPI",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/example": Object {
        "get": Object {
          "responses": Object {},
        },
      },
    },
  },
  "changelog": Array [],
  "next": Object {
    "info": Object {
      "title": "OpenAPI",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/example": Object {
        "get": Object {
          "responses": Object {},
        },
      },
    },
    "x-snyk-api-stability": "published",
  },
  "results": Array [
    Object {
      "change": Object {
        "changed": Object {
          "after": "published",
          "before": undefined,
        },
        "location": Object {
          "conceptualLocation": Object {
            "method": "",
            "path": "Resource Document",
          },
          "conceptualPath": Array [
            "x-snyk-api-stability",
          ],
          "jsonPath": "/x-snyk-api-stability",
          "kind": "SnykStability",
        },
      },
      "condition": "be provided for every resource document",
      "docsLink": "https://github.com/snyk/sweater-comb/blob/main/docs/version.md#stability-levels",
      "error": "published must be one of allowed values wip,experimental,beta,ga: expected false to be truthy",
      "isMust": true,
      "isShould": false,
      "passed": false,
      "where": "published stability: ",
    },
  ],
}
`;

exports[`lifecycle stability must be provided 1`] = `
Object {
  "base": Object {
    "info": Object {
      "title": "OpenAPI",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/example": Object {
        "get": Object {
          "responses": Object {},
        },
      },
    },
  },
  "changelog": Array [],
  "next": Object {
    "info": Object {
      "title": "OpenAPI",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/example": Object {
        "get": Object {
          "responses": Object {},
        },
      },
    },
  },
  "results": Array [
    Object {
      "change": Object {
        "changed": Object {
          "after": undefined,
          "before": undefined,
        },
        "location": Object {
          "conceptualLocation": Object {
            "method": "",
            "path": "Resource Document",
          },
          "conceptualPath": Array [
            "x-snyk-api-stability",
          ],
          "jsonPath": "/x-snyk-api-stability",
          "kind": "SnykStability",
        },
      },
      "condition": "be provided for every resource document",
      "docsLink": "https://github.com/snyk/sweater-comb/blob/main/docs/version.md#stability-levels",
      "error": "undefined must be one of allowed values wip,experimental,beta,ga: expected false to be truthy",
      "isMust": true,
      "isShould": false,
      "passed": false,
      "where": "published stability: ",
    },
  ],
}
`;

exports[`lifecycle stability wip can change to another maturity 1`] = `
Object {
  "base": Object {
    "info": Object {
      "title": "OpenAPI",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/example": Object {
        "get": Object {
          "responses": Object {},
        },
      },
    },
    "x-snyk-api-stability": "wip",
  },
  "changelog": Array [],
  "next": Object {
    "info": Object {
      "title": "OpenAPI",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/example": Object {
        "get": Object {
          "responses": Object {},
        },
      },
    },
    "x-snyk-api-stability": "beta",
  },
  "results": Array [
    Object {
      "change": Object {
        "changed": Object {
          "after": "beta",
          "before": "wip",
        },
        "location": Object {
          "conceptualLocation": Object {
            "method": "",
            "path": "Resource Document",
          },
          "conceptualPath": Array [
            "x-snyk-api-stability",
          ],
          "jsonPath": "/x-snyk-api-stability",
          "kind": "SnykStability",
        },
      },
      "condition": "not change unless it was wip",
      "docsLink": "https://github.com/snyk/sweater-comb/blob/main/docs/version.md#promoting-stability-of-a-resource-over-time",
      "isMust": true,
      "isShould": false,
      "passed": true,
      "where": "published stability: ",
    },
  ],
}
`;

exports[`lifecycle sunset fails when the file was removed and not deprecated 1`] = `
Object {
  "base": Object {
    "info": Object {
      "title": "OpenAPI",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/example": Object {
        "get": Object {
          "responses": Object {},
        },
      },
    },
  },
  "changelog": Array [],
  "next": Object {
    "info": Object {
      "title": "Empty",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/example": Object {
        "get": Object {
          "responses": Object {},
        },
      },
    },
  },
  "results": Array [
    Object {
      "change": Object {
        "added": Object {
          "changeDate": "2021-10-10",
          "changeResource": "Example",
          "changeVersion": Object {
            "date": "2021-10-10",
            "stability": "ga",
          },
          "resourceVersions": Object {},
        },
        "changeType": "added",
        "location": Object {
          "conceptualLocation": Object {
            "method": "",
            "path": "Entire Resource",
          },
          "conceptualPath": Array [],
          "jsonPath": "/",
          "kind": "ContextRule",
        },
      },
      "condition": "follow sunset rules",
      "docsLink": undefined,
      "error": "expected Example to be deprecated before removing",
      "isMust": true,
      "isShould": false,
      "passed": false,
      "where": "api lifeycle: ",
    },
  ],
}
`;

exports[`lifecycle sunset sunset schedule fails when the schedule isn't met 1`] = `
Object {
  "base": Object {
    "info": Object {
      "title": "OpenAPI",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/example": Object {
        "get": Object {
          "responses": Object {},
        },
      },
    },
  },
  "changelog": Array [],
  "next": Object {
    "info": Object {
      "title": "Empty",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/example": Object {
        "get": Object {
          "responses": Object {},
        },
      },
    },
  },
  "results": Array [
    Object {
      "change": Object {
        "added": Object {
          "changeDate": "2021-10-20",
          "changeResource": "Example",
          "changeVersion": Object {
            "date": "2021-10-10",
            "stability": "experimental",
          },
          "resourceVersions": Object {
            "Example": Object {
              "2021-10-10": Object {
                "experimental": Object {
                  "deprecatedBy": Object {
                    "date": "2021-10-20",
                    "stability": "ga",
                  },
                },
              },
            },
          },
        },
        "changeType": "added",
        "location": Object {
          "conceptualLocation": Object {
            "method": "",
            "path": "Entire Resource",
          },
          "conceptualPath": Array [],
          "jsonPath": "/",
          "kind": "ContextRule",
        },
      },
      "condition": "follow sunset rules",
      "docsLink": undefined,
      "error": "expected experimental resource Example to be deprecated 30 days: expected 10 to be above 30",
      "isMust": true,
      "isShould": false,
      "passed": false,
      "where": "api lifeycle: ",
    },
  ],
}
`;

exports[`lifecycle sunset sunset schedule passes when the schedule is met 1`] = `
Object {
  "base": Object {
    "info": Object {
      "title": "OpenAPI",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/example": Object {
        "get": Object {
          "responses": Object {},
        },
      },
    },
  },
  "changelog": Array [],
  "next": Object {
    "info": Object {
      "title": "Empty",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/example": Object {
        "get": Object {
          "responses": Object {},
        },
      },
    },
  },
  "results": Array [
    Object {
      "change": Object {
        "added": Object {
          "changeDate": "2021-12-01",
          "changeResource": "Example",
          "changeVersion": Object {
            "date": "2021-10-10",
            "stability": "experimental",
          },
          "resourceVersions": Object {
            "Example": Object {
              "2021-10-10": Object {
                "experimental": Object {
                  "deprecatedBy": Object {
                    "date": "2021-10-20",
                    "stability": "ga",
                  },
                },
              },
            },
          },
        },
        "changeType": "added",
        "location": Object {
          "conceptualLocation": Object {
            "method": "",
            "path": "Entire Resource",
          },
          "conceptualPath": Array [],
          "jsonPath": "/",
          "kind": "ContextRule",
        },
      },
      "condition": "follow sunset rules",
      "docsLink": undefined,
      "isMust": true,
      "isShould": false,
      "passed": true,
      "where": "api lifeycle: ",
    },
  ],
}
`;
